src/rockstor/storageadmin/views/share_helpers.py

```diff
@@ -78,20 +80,48 @@ def import_shares(pool, request):
     # Find the actual/current shares/subvols within the given pool:
     # Limited to Rockstor relevant subvols ie shares and clones.
     shares_in_pool = shares_info(pool)
+    # List of pool's share.pqgroups so we can remove inadvertent duplication.
+    # All pqgroups are removed when quotas are disabled, combined with a part
+    # refresh we could have duplicates within the db.
+    share_pqgroups_used = []
     # Delete db Share object if it is no longer found on disk.
     for s_in_pool_db in shares_in_pool_db:
         if s_in_pool_db not in shares_in_pool:
             Share.objects.get(pool=pool, name=s_in_pool_db).delete()
     # Check if each share in pool also has a db counterpart.
     for s_in_pool in shares_in_pool:
-        logger.debug('Share name = {}.'.format(s_in_pool))
+        logger.debug('---- Share name = {}.'.format(s_in_pool))
         if s_in_pool in shares_in_pool_db:
             logger.debug('Updating pre-existing same pool db share entry.')
             # We have a pool db share counterpart so retrieve and update it.
             share = Share.objects.get(name=s_in_pool, pool=pool)
+            # Initially default our pqgroup value to db default of '-1/-1'
+            # This way, unless quotas are enabled, all pqgroups will be
+            # returned to db default.
+            pqgroup = settings.MODEL_DEFS['pqgroup']
+            if share.pool.quotas_enabled:
+                # Quotas are enabled on our pool so we can validate pqgroup.
+                if share.pqgroup == pqgroup or not share.pqgroup_exist \
+                        or share.pqgroup in share_pqgroups_used:
+                    # we have a void '-1/-1' or non existent pqgroup or
+                    # this pqgroup has already been seen / used in this pool.
+                    logger.debug('#### replacing void, non-existent, or '
+                                 'duplicate pqgroup')
+                    pqgroup = qgroup_create(pool)
+                    update_quota(pool, pqgroup, share.size * 1024)
+                    share_pqgroup_assign(pqgroup, share)
+                else:
+                    # Our share's pqgroup looks OK so use it.
+                    pqgroup = share.pqgroup
+                # Record our use of this pqgroup to spot duplicates later.
+                share_pqgroups_used.append(deepcopy(share.pqgroup))
```

Comment:
I am unable to figure out why we need a deepcopy here. Could you add a comment explaining why?

Reply:
I don't have one worth of code comment but I can comment here that I was just playing safe: ie avoiding any accidental same object dual reference type thing and never revisited it. Your call on if absolutely required. I know they are expensive but just didn't want to risk accidental dual reference and have our taken copy still referencing the source, ie proper separation. I did all my testing of expected behaviour with the deepcopy in place.

Hope that helps on the why.